src/rust/engine/process_execution/src/remote.rs

```diff
@@ -685,8 +713,16 @@ impl CommandRunner {
         Err(err) => match err {
           ExecutionError::Fatal(e) => return Err(e),
           ExecutionError::Retryable(e) => {
-            // do nothing, will retry
+            // Check if the number of request attempts sent thus far have exceeded the number
+            // of retries allowed since the last successful connection. (There is no point in
+            // continually submitting a request if ultimately futile.)
             trace!("retryable error: {}", e);
+            if num_retries >= MAX_RETRIES {
+              return Err(format!("Too many failures from server, last error: {}", e));
+            } else {
+              // Increment the retry counter and allow loop to retry.
+              num_retries += 1;
```

Comment:
I'm a little confused why `num_retries` is being used in two different blocks. Should there be two separate variables, or it does make sense to use the same variable in both contexts?

Reply:
Those two blocks are the only two paths that end up back at the top of the loop. It was easier to just duplicate the code, rather than figure out a way to have them share that code. Otherwise I'd have to do tricks (like a `mut` variable) to store the error away for this code being at the top of the loop.

Reply:
To clarify, I'm not concerned with DRY. I'm concerned if it makes sense for the first code block having to have retried impacting the second block? Although I think I misread the code - it looks like these two blocks are mutually exclusive? We're either using the Stream API or the other API (poll API?)? We'll never use both at the same time?

Reply:
Both `Execute` and `WaitExecution` methods are streaming APIs. We do not poll the Operations API any more.

One block handles the server just disconnecting; the other handles the server reporting back with a retryable error. In both cases, we should retry but limit the number of retries.